Add blog post announcing Oracle GraalVM support (#71)
paketo-buildpacks/blog: content/posts/0025-oracle-graalvm-support.md

---
title: Oracle GraalVM Support has Arrived!
date: "2023-12-11"
slug: oracle-graalvm-support-has-arrived
author: dmikusa
---

## Oracle GraalVM Support has Arrived!

Today marks the start of support for building native image applications with [Oracle GraalVM](https://www.oracle.com/java/graalvm/). Since Oracle released Oracle GraalVM under the [GraalVM Free License](https://blogs.oracle.com/java/post/graalvm-free-license), we've been receiving lots of feedback to add support for it. Starting with the 3.11.0 release of the Paketo Buildpack for Oracle we now have official support! 

With this first release, you can build native image applications with Oracle GraalVM. Building your application is simple. You follow the [process for selecting a Native Image Toolkit](https://paketo.io/docs/howto/java/#use-an-alternative-java-native-image-toolkit) and use the Oracle buildpack. 

> For example, if we want to build the [Paketo Maven Spring Boot Sample App](https://github.com/paketo-buildpacks/samples/tree/main/java/native-image/spring-boot-native-image-maven) with `pack`, we would simply run `pack build applications/native-image -b paketo-buildpacks/oracle -b urn:cnb:builder:paketo-buildpacks/java-native-image --builder paketobuildpacks/builder-jammy-tiny --env BP_MAVEN_ACTIVE_PROFILES=native`. This small change is all that's required to tell Paketo Buildpacks to override the default Native Image Toolkit, which is from Bellsoft, and instead use Oracle GraalVM.

Reviewer: ```suggestion
For example, if we want to build the [Paketo Maven Spring Boot Sample App](https://github.com/paketo-buildpacks/samples/tree/main/java/native-image/spring-boot-native-image-maven) with `pack`, we would simply run `pack build applications/native-image -b paketo-buildpacks/oracle -b urn:cnb:builder:paketo-buildpacks/java-native-image --builder paketobuildpacks/builder-jammy-tiny --env BP_MAVEN_ACTIVE_PROFILES=native`. This small change is all that's required to tell Paketo Buildpacks to override the default Native Image Kit (NIK), which is from Bellsoft, and instead use Oracle GraalVM.
```

> For example, if we want to build the [Paketo Maven Spring Boot Sample App](https://github.com/paketo-buildpacks/samples/tree/main/java/native-image/spring-boot-native-image-maven) with `pack`, we would simply run `pack build applications/native-image -b paketo-buildpacks/oracle -b urn:cnb:builder:paketo-buildpacks/java-native-image --builder paketobuildpacks/builder-jammy-tiny --env BP_MAVEN_ACTIVE_PROFILES=native`. This small change is all that's required to tell Paketo Buildpacks to override the default Native Image Toolkit, which is from Bellsoft, and instead use Oracle GraalVM.

Reviewer 2: No problem! I admit it can get confusing at times. ;)

Just as easily, you can swap versions too. The Oracle buildpack presently supports both Java 17 and 21, which you can select by setting the `BP_JVM_VERSION` flag. For example, `pack build applications/native-image -b paketobuildpacks/oracle -b urn:cnb:builder:paketo-buildpacks/java-native-image --builder paketobuildpacks/builder-jammy-tiny --env BP_MAVEN_ACTIVE_PROFILES=native --env BP_JVM_VERSION=21`. This makes trying out new versions trivial, which ultimately makes your upgrades easier.

### Spring Boot Build Tools

If you're using Spring Boot Build Tools to generate your application image, you can use Oracle GraalVM with the following changes.

1. Modify the image configuration block in your `pom.xml`.

    ```
    ...
    <configuration>
        <image>
            <builder>paketobuildpacks/builder-jammy-tiny:latest</builder>
            <buildpacks>
                <buildpack>docker.io/paketobuildpacks/oracle</buildpack>
                <buildpack>urn:cnb:builder:paketo-buildpacks/java-native-image</buildpack>
            </buildpacks>
        </image>
    </configuration>
    ...
    ```

2. Then run `./mvnw spring-boot:build-image -Pnative` to build your image.

> 

Reviewer: ```suggestion

And the equivalent change with Gradle `build.gradle`:
    ```
    ...
    tasks.named("bootBuildImage") {
        environment = ['BP_NATIVE_IMAGE': 'true']
        builder = "paketobuildpacks/builder-jammy-tiny:latest"
        buildpacks = ["docker.io/paketobuildpacks/oracle", "urn:cnb:builder:paketo-buildpacks/java-native-image"]
    }
    ...
    ```
2. Then run `./gradlew bootBuildImage` to build your image.

```

Reviewer: it was already set in https://github.com/paketo-buildpacks/samples/blob/main/java/native-image/spring-boot-native-image-gradle/build.gradle#L18 ; maybe it's no longer necessary...

Reviewer: so I retried with this env. variable, and it worked fine.
Thanks for spotting this simplication!

## Future Roadmap

At the moment, the Oracle buildpack supports using the [Oracle free JDK](https://www.oracle.com/java/technologies/downloads/) for running your Java apps, and it supports using [Oracle GraalVM](https://www.oracle.com/java/graalvm/) for building your Native Image apps. What's been left out at the moment is support to use Oracle GraalVM as a JDK to run your Java apps.

The was something we left out of the initial support because it would have delayed the release more and that's something we didn't want to do. Also, we beleive that future work in the Paketo project is going to make support for this significantly easier.

What this all means is that for the moment, you can't run your Java apps with GraalVM. We'll be working hard to support this in the future, but in meantime we think you'll really enjoy using Oracle GraalVM to build your native images.

> As always, please post questions/comments on our [Github Discussion Page](https://github.com/orgs/paketo-buildpacks/discussions/categories/java-team), or Join the [Paketo Slack](https://slack.paketo.io/) and chat with us in the `#java` room.

Reviewer 2: ```suggestion
As always, please post questions/comments on our [GitHub Discussion Page](https://github.com/orgs/paketo-buildpacks/discussions/categories/java-team), or Join the [Paketo Slack](https://slack.paketo.io/) and chat with us in the `#java` room.
```
